compiler/grainformat/comment_utils.re

```diff
@@ -0,0 +1,564 @@
+          Locations.get_raw_pos_info(
+            Locations.get_comment_loc(cmt).loc_start,
+          );
+
+        if (bracket_line == this_line) {
+          [
+            Doc.space,
+            comment_to_doc(cmt),
+            ...comments_inner(Some(cmt), bracket_line_opt, rem),
+          ];
+        } else {
+          [
+            Doc.hardLine,
+            comment_to_doc(cmt),
+            ...comments_inner(Some(cmt), bracket_line_opt, rem),
+          ];
+        };
+      }
+    }
+  | Some(prev_cmt) =>
+    switch (comments) {
+    | [] => [Doc.nil]
+    | [cmt, ...rem] =>
+      let (_, prev_line, _, _) =
+        Locations.get_raw_pos_info(
+          Locations.get_comment_loc(prev_cmt).loc_end,
+        );
+      let (_, this_line, _, _) =
+        Locations.get_raw_pos_info(Locations.get_comment_loc(cmt).loc_start);
+
+      switch (this_line - prev_line) {
+      | 0 => [
+          Doc.space,
+          comment_to_doc(cmt),
+          ...comments_inner(Some(cmt), bracket_line_opt, rem),
+        ]
+
+      | 1 => [
+          switch (prev_cmt) {
+          | Line(_) => Doc.nil
+          | _ => Doc.hardLine
+          },
+          comment_to_doc(cmt),
+          ...comments_inner(Some(cmt), bracket_line_opt, rem),
+        ]
+      | _ => [
+          switch (prev_cmt) {
+          | Doc(_)
+          | Shebang(_)
+          | Line(_) => Doc.nil
+          | _ => Doc.hardLine
+          },
+          Doc.hardLine,
+          comment_to_doc(cmt),
+          ...comments_inner(Some(cmt), bracket_line_opt, rem),
+        ]
+      };
+    }
+  };
+};
+
+let inbetween_comments_to_docs =
+    (
+      ~offset: bool,
+      ~bracket_line: option(int),
+      comments: list(Parsetree.comment),
+    ) =>
+  switch (comments) {
+  | [] =>
+    switch (bracket_line) {
+    | None => Doc.nil
+    | Some(_) => Doc.softLine
+    }
+  | _remaining_comments =>
+    if (offset) {
+      Doc.concat([
+        Doc.space,
+        Doc.concat(comments_inner(None, bracket_line, comments)),
+      ]);
+    } else {
+      Doc.concat(comments_inner(None, bracket_line, comments));
+    }
+  };
+
+let rec get_comments_after_location =
+        (~location: Location.t, comments: list(Parsetree.comment)) => {
+  let (_, stmt_end_line, stmt_end_char, _) =
+    Locations.get_raw_pos_info(location.loc_end);
+
+  switch (comments) {
+  | [] => []
+  | [cmt, ...remaining_comments] =>
+    let c_loc: Location.t = Locations.get_comment_loc(cmt);
+
+    let (_, cmtsline, cmtschar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_start);
+
+    if (cmtsline > stmt_end_line
+        || cmtsline == stmt_end_line
+        && cmtschar > stmt_end_char) {
+      [cmt, ...get_comments_after_location(~location, remaining_comments)];
+    } else {
+      get_comments_after_location(~location, remaining_comments);
+    };
+  };
+};
+
+let rec trailing_comments_inner =
+        (prev: option(Parsetree.comment), comments: list(Parsetree.comment)) => {
```

Comment:
Why isn't this an optional, named param?

Reply:
Nice, that's changed.